llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp

```diff
@@ -527,6 +527,14 @@ DIE &DwarfCompileUnit::updateSubprogramScopeDIE(const DISubprogram *SP) {
           *DD->getCurrentFunction()))
     addFlag(*SPDie, dwarf::DW_AT_APPLE_omit_frame_ptr);
 
+  if (Asm->OutStreamer->getGenerateFuncLineTableOffsets() &&
+      Asm->OutStreamer->getCurrentFuncFirstLineStreamSym()) {
+    addSectionLabel(
+        *SPDie, dwarf::DW_AT_LLVM_stmt_sequence,
+        Asm->OutStreamer->getCurrentFuncFirstLineStreamSym(),
+        Asm->getObjFileLowering().getDwarfLineSection()->getBeginSymbol());
+  }
```

Comment:
Rather than wiring up the attribute into the streamer, then querying it out here in DwarfCompileUnit, then going back into the streamer with labels - could the streamer "do the right thing" when a label is requested, and otherwise do the old/usual thing?

It doesn't seem like MCStreamer should "know" about function labels, it should know about line table labels, and you could in theory request them anywhere you want to be able to jump into parsing the line table without having to parse additional context. (ie: in `beginFunctionImpl` ? (though maybe that's after the prologue, in which case maybe there's some callback before the prologue, or we should add one))

Reply:
That sounds like a better design and I think it may work. For the current implementation - same as above - was trying to follow Apple's existing design. See [here](https://github.com/swiftlang/llvm-project/blob/601abd74df1b6cad0690397898a2bd3a27a955e1/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp#L2374) for example. 

Before taking the more straightforward approach just want to make sure we aren't necessarily weighing in Apple's existing design - and we should try to get the best design for the current feature only.

Reply:
Ah, fair enough -prior art and all.

Could you rope someone in from that work to join this discussion, then?

Reply:
@rastogishubham / @adrian-prantl - In this PR I'm doing something similar to `-fcas-friendly-debug-info` in the Apple branch (and also a bit extra on top). Do you think this is the way to go for future merges / consistency / etc ?

Reply:
@bulbazord / @kastiglione - We're trying to do something similar to `-fcas-friendly-debug-info` in the Apple branch - could you have a look if you're OK with this approach for consistency with Apple branch ?

Reply:
Sorry, I missed this when you pinged me originally.
There is no good reason for `-fcas-friendly-debug-info` to live downstream so we should try to unify both implementations.
@rastogishubham Can you take a look at both patches and see if we could rebase `-fcas-friendly-debug-info` on top of this?

Reply:
Thanks for the ping, let me take a look, I need to also familiarize myself with `-fcas-friendly-debug-info` code again

Reply:
This feedback still seems relevant, to me, to the current version of the code - it's still strange to me that MCStreamer needs to know about this with an attribute up-front with 'setGenerateFuncLineTableOffsets` - in assembly that isn't necessary, the asm directive is used wherever needed, and MCStreamer behaves appropriately/as needed (it's not like we needed a directive at the start of the file to announce that we might use line table offset directives later in the file - so why would we need that at the API level? We should just query for line table offsets when we need them and MCStreamer should provide them at that point, without needing some prior warning/up front thing)

Reply:
Good call. The new implementation seems much more straightforward. LMK what you think. 

I also tested it on [a more complex example](https://gist.github.com/alx32/f35023b7b2939af4d6110f8c37f4d918) and [the result](https://www.diffchecker.com/QesUhKWy/) seems correct.